fairworkflows/fairstep.py

```diff
@@ -334,3 +355,78 @@ def __str__(self):
         s = f'Step URI = {self._uri}\n'
         s += self._rdf.serialize(format='trig').decode('utf-8')
         return s
+
+
+def mark_as_fairstep(label: str = None, is_pplan_step: bool = True, is_manual_task: bool = None,
+                     is_script_task: bool = None):
+    """Mark a function as a FAIR step.
+
+    Use as decorator to mark a function as a FAIR step. Set properties of the fair step using
+    arguments to the decorator.
+
+    The raw code of the function will be used to set the description of the fair step.
+
+    The type annotations of the input arguments and return statement will be used to
+    automatically set inputs and outputs of the FAIR step.
+
+    Args:
+        label (str): Label of the fair step (corresponds to rdfs:label predicate)
+        is_pplan_step (str): Denotes whether this step is a pplan:Step
+        is_manual_task (str): Denotes whether this step is a bpmn.ManualTask
+        is_script_task (str): Denotes whether this step is a bpmn.ScriptTask
+
+    """
+    def _modify_function(func):
+        """
+        Store FairStep object as _fairstep attribute of the function. Use inspection to get the
+        description, inputs, and outputs of the step based on the function specification.
+        """
+        def _wrapper(*args, **kwargs):
+            return func(*args, **kwargs)
+        # Description of step is the raw function code
+        description = inspect.getsource(func)
+        inputs = _extract_inputs_from_function(func)
+        outputs = _extract_outputs_from_function(func)
+        _wrapper._fairstep = FairStep(label=label,
+                                      description=description,
+                                      is_pplan_step=is_pplan_step,
+                                      is_manual_task=is_manual_task,
+                                      is_script_task=is_script_task,
+                                      inputs=inputs,
+                                      outputs=outputs)
+        _wrapper._fairstep.validate()
+        return _wrapper
+    return _modify_function
+
+
+def _extract_inputs_from_function(func) -> List[FairVariable]:
+    """
+    Extract inputs from function using inspection. The name of the argument will be the name of
+    the fair variable, the corresponding type hint will be the type of the variable.
+    """
+    argspec = inspect.getfullargspec(func)
+    try:
+        return [FairVariable(name=arg, type=argspec.annotations[arg].__name__)
+                for arg in argspec.args]
+    except KeyError:
+        raise ValueError('Not all input arguments have type hinting,'
```

Comment:
I think we should explicitly state that 'Fair' steps _must_ have type hinting, in this error message. I can see some people being slightly confused otherwise about what the error is here, unless they realise it's a design choice we've made.

Reply:
Very good point! I addressed it :)